File: examples/typescript/05_merchant_checkout.ts
// Scenario 14: a minimal merchant checkout backend.
// Demonstrates the SDK building blocks + an in-memory order store. In production
// use a real database, a unique address per order, and signed webhooks.
import {
  KingPepeClient,
  createPaymentRequest,
  paymentUri,
  checkPayment,
  parseKpepe,
  toPaymentEvent,
  type PaymentRequest,
  type PaymentStatus,
} from "@kingpepe/sdk";

interface Order {
  request: PaymentRequest;
  status: PaymentStatus;
  fulfilled: boolean;
}

const orders = new Map<string, Order>(); // orderId -> Order (use a DB in production)

export async function createCheckout(wallet: KingPepeClient, orderId: string, priceKpepe: string) {
  const request = createPaymentRequest({
    id: orderId,
    address: await wallet.getNewAddress(orderId), // unique address per order
    amountBaseUnits: parseKpepe(priceKpepe),
    requiredConfirmations: 2,
    ttlMs: 20 * 60 * 1000,
  });
  orders.set(orderId, { request, status: "pending", fulfilled: false });
  return { orderId, uri: paymentUri(request), address: request.address };
}

// Call this on a timer / from a block notification. Idempotent: fulfils once.
export async function pollOrder(wallet: KingPepeClient, orderId: string, emit: (e: unknown) => void) {
  const order = orders.get(orderId);
  if (!order || order.fulfilled) return;
  const state = await checkPayment(wallet, order.request);
  if (state.status !== order.status) {
    order.status = state.status;
    emit(toPaymentEvent(state)); // send a signed webhook in production
  }
  if (state.status === "paid" && !order.fulfilled) {
    order.fulfilled = true; // deliver the product exactly once
    console.log(`Order ${orderId} PAID — fulfilling.`);
  }
}

async function demo() {
  const wallet = KingPepeClient.fromEnv().withWallet("example-shop");
  const checkout = await createCheckout(wallet, `order-${Date.now()}`, "3.00");
  console.log("Checkout:", checkout);
  await pollOrder(wallet, checkout.orderId, (e) => console.log("webhook:", JSON.stringify(e)));
}

if (process.argv[1] && process.argv[1].endsWith("05_merchant_checkout.ts")) {
  demo().catch((err) => {
    console.error("error:", err.message);
    process.exit(1);
  });
}

File: examples/python/05_merchant_checkout.py
"""Scenario 14: a minimal merchant checkout backend.

Demonstrates the SDK building blocks + an in-memory order store. In production
use a real database, a unique address per order, and signed webhooks.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Callable

from kingpepe_sdk import (
    KingPepeClient,
    PaymentRequest,
    check_payment,
    create_payment_request,
    parse_kpepe,
    payment_uri,
    to_payment_event,
)


@dataclass
class Order:
    request: PaymentRequest
    status: str = "pending"
    fulfilled: bool = False


ORDERS: dict[str, Order] = {}  # order_id -> Order (use a DB in production)


def create_checkout(wallet: KingPepeClient, order_id: str, price_kpepe: str) -> dict:
    request = create_payment_request(
        id=order_id,
        address=wallet.get_new_address(order_id),  # unique address per order
        amount_base_units=parse_kpepe(price_kpepe),
        required_confirmations=2,
        ttl_ms=20 * 60 * 1000,
    )
    ORDERS[order_id] = Order(request=request)
    return {"order_id": order_id, "uri": payment_uri(request), "address": request.address}


def poll_order(wallet: KingPepeClient, order_id: str, emit: Callable[[dict], None]) -> None:
    """Idempotent: safe to call repeatedly; fulfils the order exactly once."""
    order = ORDERS.get(order_id)
    if order is None or order.fulfilled:
        return
    state = check_payment(wallet, order.request)
    if state.status != order.status:
        order.status = state.status
        emit(to_payment_event(state))  # send a signed webhook in production
    if state.status == "paid" and not order.fulfilled:
        order.fulfilled = True
        print(f"Order {order_id} PAID — fulfilling.")


def _demo() -> None:
    wallet = KingPepeClient.from_env().with_wallet("example-shop")
    checkout = create_checkout(wallet, f"order-{int(time.time())}", "3.00")
    print("Checkout:", checkout)
    poll_order(wallet, checkout["order_id"], lambda e: print("webhook:", e))


if __name__ == "__main__":
    _demo()
